hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/TestHoodieClientMultiWriter.java

```diff
@@ -545,6 +546,120 @@ public void testMultiWriterWithAsyncTableServicesWithConflict(HoodieTableType ta
+
+    // Three clients running actions in parallel
+    final int threadCount = 3;
+    final ExecutorService executor = Executors.newFixedThreadPool(threadCount);
+
+    final SparkRDDWriteClient client1 = getHoodieWriteClient(cfg);
+    final SparkRDDWriteClient client2 = getHoodieWriteClient(cfg);
+    final String commitTime2 = "002";
+    final String commitTime3 = "003";
+    AtomicReference<Object> writeStatus1 = new AtomicReference<>(null);
+    AtomicReference<Object> writeStatus2 = new AtomicReference<>(null);
+
+    Future future1 = executor.submit(() -> {
+      final int numRecords = 100;
+      assertDoesNotThrow(() -> {
+        writeStatus1.set(createCommitWithInserts(cfg, client1, "001", commitTime2, numRecords, false));
+      });
+    });
+    Future future2 = executor.submit(() -> {
+      final int numRecords = 100;
+      assertDoesNotThrow(() -> {
+        writeStatus2.set(createCommitWithInserts(cfg, client2, "001", commitTime3, numRecords, false));
+      });
+    });
+
+    future1.get();
+    future2.get();
+
+    final CountDownLatch commitCountDownLatch = new CountDownLatch(2);
+    HoodieTableMetaClient tableMetaClient = client.getTableServiceClient().createMetaClient(true);
+
+    // Get inflight instant stream before commit
+    List<HoodieInstant> inflightInstants = client
+            .getTableServiceClient()
+            .getInflightTimelineExcludeCompactionAndClustering(tableMetaClient)
+            .getReverseOrderedInstants()
+            .collect(Collectors.toList());
+
+    // Commit the instants and get instants to rollback in parallel
+    future1 = executor.submit(() -> {
+      long start = System.currentTimeMillis();
+      LOG.info(String.format("Start to commit instant %s", commitTime2));
+      client1.commit(commitTime2, writeStatus1.get());
+      commitCountDownLatch.countDown();
+      validInstants.add(commitTime2);
+      LOG.info(String.format("commit the instant cost %d ms", System.currentTimeMillis() - start));
+    });
+
+    future2 = executor.submit(() -> {
+      long start = System.currentTimeMillis();
+      LOG.info(String.format("Start to commit instant %s", commitTime3));
+      client2.commit(commitTime3, writeStatus2.get());
+      commitCountDownLatch.countDown();
+      validInstants.add(commitTime3);
+      LOG.info(String.format("commit the instant %s cost %d ms", commitTime3, System.currentTimeMillis() - start));
+    });
+
+    Future future3 = executor.submit(() -> {
+      try {
+        commitCountDownLatch.await(30000, TimeUnit.MILLISECONDS);
+      } catch (InterruptedException e) {
+        //
+      }
+      LOG.info("Start to get instants to rollback");
+      List<String> instantsToRollback =
+              client.getTableServiceClient().getInstantsToRollbackForLazyCleanPolicy(tableMetaClient, inflightInstants.stream());
+      // No instants will be rollback though some instants may be detected as `expired`
+      assertTrue(instantsToRollback.isEmpty());
```

Comment:
why not make commit2 inflight so we have 1 instant to return here while having commit3 completed and not included? this will give a better coverage

Reply:
Nice suggestions! will modify it